### ISSUE TITLE
Generate schemas automatically at build time

### ISSUE THREAD

**Issue by mikavilpas** [yazi-rs/schemas#24]
I'm not sure how this could fit the existing ecosystem, but bear with me. 🙂 I looked over the currently open issues and there seems to be some issues with versioning as well as the schemas falling behind the newest changes. Maybe this could be improved a bit.

I would like to propose an approach where the schemas are generated automatically. A while ago I helped contribute a similar change in https://github.com/mfontanini/presenterm/pull/228 and https://github.com/mfontanini/presenterm/pull/236. The maintainer proposed doing it using https://docs.rs/schemars/latest/schemars/ , and it worked nicely.

**Comment by mikavilpas** [comment]
Oh, sorry if it came off badly - I have had 0 issues with yazi's schemas myself. I no longer remember what I meant, but maybe it's not that important.

If the current way of updating works - great 👍🏻. If this suggestion seems worth trying, great too 🙂 

> I don't think this solves our versioning issues?

I think you're correct - one schema applies for one version as far as I am aware.

Right now I think yazi has a latest stable release as well as a constantly changing bleeding edge. As long as the latest release is bound to either a git tag or branch, the schema's url in the configuration file can point to that and it should stay up to date.

> not sure how complex of conditions it can handle

Hmm this is a good point. The schemas I worked on for presenterm were quite simple. Maybe a small proof-of-concept could be a good idea.

**Comment by uncenter** [comment]
> > I don't think this solves our versioning issues?
> 
> I think you're correct - one schema applies for one version as far as I am aware.
> 
> Right now I think yazi has a latest stable release as well as a constantly changing bleeding edge. As long as the latest release is bound to either a git tag or branch, the schema's url in the configuration file can point to that and it should stay up to date.

I believe I mentioned branching/tagging in #17 but for some reason we didn't land on it. Anyway, sure yeah that sounds doable.

> > not sure how complex of conditions it can handle
> 
> Hmm this is a good point. The schemas I worked on for presenterm were quite simple. Maybe a small proof-of-concept could be a good idea.

Definitely, would appreciate that!

**Comment by uncenter** [comment]
> The maintainer proposed doing it using [docs.rs/schemars/latest/schemars](https://docs.rs/schemars/latest/schemars/) , and it worked nicely.

@sxyazi I wonder if we should look into autogenerating the schemas in the main Yazi repository again with schemars. Might make things easier!

**Comment by sxyazi** [comment]
Maybe generating from docs is easier, since we already have [typegen.js](https://github.com/yazi-rs/yazi-rs.github.io/blob/main/scripts/typegen.js) which generates [Lua type definitions](https://github.com/yazi-rs/plugins/tree/main/types.yazi), and perhaps we can extend its output to schemas as well, and the doc-parsing part is in place.

We also need to add type annotations for each property in [yazi.toml](https://yazi-rs.github.io/docs/configuration/yazi) just like what [theme.toml](https://yazi-rs.github.io/docs/configuration/theme) did so that `typegen.js` can use those types to produce schemas, but I'm not sure how to represent `or` or `and` in the schemas yet.

**Comment by uncenter** [comment]
I don't think that is easier, I think we should definitely be using schemars. I think that is the only way to represent the complexity of types without reimplemtning schemars itself haha. Additionally, I feel like generating documentation and schemas and types from the Rust code is easier than generating schemas and types from the documentation? Especially given the power of schemars.
